sonic_installer/main.py

```diff
@@ -317,5 +317,68 @@ def cleanup():
     if image_removed == 0:
         click.echo("No image(s) to remove")
 
+# Upgrade docker image
+@cli.command()
+@click.option('-y', '--yes', is_flag=True, callback=abort_if_false,
+        expose_value=False, prompt='New docker image will be installed, continue?')
+@click.option('--cleanup_image', is_flag=True, help="Clean up old docker images")
+@click.argument('container_name', metavar='<container_name>', required=True, type=click.Choice(["swss", "snmp", "lldp"]))
+@click.argument('tag', metavar='<tag>', required=True, type=click.STRING)
+@click.argument('url')
+def upgrade_docker(cleanup_image, container_name, tag, url):
+    """ Upgrade docker image from local binary or URL"""
+
+    # example images: docker-lldp-sv2:latest
+    cmd = "docker inspect --format '{{.Config.Image}}' " + container_name
+    proc = subprocess.Popen(cmd, stdout=subprocess.PIPE, shell=True)
+    images = proc.stdout.read().rstrip()
+
+    # example image_name: docker-lldp-sv2
+    cmd = "echo " + images + " | cut -d ':' -f 1"
+    proc = subprocess.Popen(cmd, stdout=subprocess.PIPE, shell=True)
+    image_name = proc.stdout.read().rstrip()
+
+    # example image_id: 7ed919240fb9
+    cmd = "docker images --format '{{.ID}}' " + image_name
+    proc = subprocess.Popen(cmd, stdout=subprocess.PIPE, shell=True)
+    image_id = proc.stdout.read()
+
+
+    DEFAULT_IMAGE_PATH = os.path.join("/tmp/", image_name)
+
+    if url.startswith('http://') or url.startswith('https://'):
+        click.echo('Downloading image...')
+        urllib.urlretrieve(url, DEFAULT_IMAGE_PATH, reporthook)
+        image_path = DEFAULT_IMAGE_PATH
+    else:
+        image_path = os.path.join("./", url)
+
+    # make sure orchagent is in clean state if swss is to be upgraded
+    if container_name == "swss":
+        cmd = "docker exec -it " + container_name + " orchagent_restart_check"
+        proc = subprocess.Popen(cmd, stdout=subprocess.PIPE, shell=True)
+        result = proc.stdout.read().rstrip()
+        if result != "RESTARTCHECK succeeded":
+            click.echo("Orchagent is not in clean state, check syslog and try later")
+            sys.exit(1)
+        else:
+            click.echo("Orchagent is in clean state and frozen for warm upgrade")
+
+    run_command("systemctl stop %s" % container_name)
+    run_command("docker rm  %s" % container_name)
+    run_command("docker rmi  %s" % images)
+    run_command("docker load < %s" % image_path)
+    run_command("docker tag %s:latest %s:%s" % (image_name, image_name, tag))
```

Comment:
I think we should tag the docker image during the building process.  By default, at upgrade docker phase, we just use that tag.  If we want to force to a different tag,  this needs to be a optional input of the CLI.

Reply:
The tag parameter could be made optional.  
Providing meta data like label to docker image and using that as tag during docker image upgrade require more work.  It might be better do it in separate PRs (multiple repos).  For now,  plan to provide function to get the default dummy tag.

Reply:
ok